### PR TITLE
power off iDisp link during D0i3

### PR DESCRIPTION
**Pull request by ranj063**

### CODE REVIEW

File: sound/pci/hda/hda_codec.c [hunk 1]
```diff
@@ -2937,32 +2938,59 @@ static void hda_call_codec_resume(struct hda_codec *codec)
 static int hda_codec_suspend(struct device *dev)
 {
 	struct hda_codec *codec = dev_to_hda_codec(dev);
+	struct hdac_device *hdev = &codec->core;
+	struct hdac_ext_link *hlink;
```
Comment (kv2019i) [inline sound/pci/hda/hda_codec.c]: I think the idea is good, but we need to formalize hda_ext usage from codec code a bit better. This now assumes controller is using the ext interface, which does not hold true for the legacy driver.

**Review by kv2019i — CHANGES_REQUESTED** [review]
First patch looks ok now. I think the second patch is the right idea, but needs work on implementation. I also check if I can come up with a cleaner approach to this today.

File: sound/pci/hda/hda_codec.c [hunk 2]
```diff
@@ -2937,32 +2938,59 @@ static void hda_call_codec_resume(struct hda_codec *codec)
...
+		dev_err(dev, "hdac link not found\n");
+		return -EIO;
+	}
+
 	codec_display_power(codec, true);
-	snd_hdac_codec_link_up(&codec->core);
+	if (codec->display_power_control)
+		snd_hdac_ext_bus_link_get(hdev->bus, hlink);
+	else
+		snd_hdac_codec_link_down(&codec->core);
```
Comment (lyakh) [inline sound/pci/hda/hda_codec.c]: `link_up()`?

**Review by lyakh — CHANGES_REQUESTED** [review]

**Comment by kv2019i** [comment]
@ranj063 I uploaded an alternative to the second patch:
https://github.com/thesofproject/linux/pull/2732

If you are ok, I can send as RFC to Takashi. This is first-ever extension of HDA extended core towards codec driver code, so this will warrant some extra review.

**Review by kv2019i — DISMISSED** [review]

**Review by plbossart — APPROVED** [review]
LGTM, nice evolution from the initial code. Thanks @ranj063 and @kv2019i

**Review by keyonjie — COMMENTED** [review]
Hi @ranj063 @kv2019i @plbossart only one concern here, is our Audio IP designed to run iDisp link with D0i3 set? If so, maybe not wise to power down dsp link blindly here, otherwise, we might need revisit this part and back and forth change could happen.

**Review by lyakh — APPROVED** [review]

**Comment by plbossart** [comment]
> Hi @ranj063 @kv2019i @plbossart only one concern here, is our Audio IP designed to run iDisp link with D0i3 set? If so, maybe not wise to power down dsp link blindly here, otherwise, we might need revisit this part and back and forth change could happen.

D0i3 is opt-in, we have to specify which streams/outputs are D0i3 compatible so we are in complete control here.

**Comment by keyonjie** [comment]
> > Hi @ranj063 @kv2019i @plbossart only one concern here, is our Audio IP designed to run iDisp link with D0i3 set? If so, maybe not wise to power down dsp link blindly here, otherwise, we might need revisit this part and back and forth change could happen.
> 
> D0i3 is opt-in, we have to specify which streams/outputs are D0i3 compatible so we are in complete control here.

Imagine that if we want to run HDMI audio on D0i3 and specify d0i3_compatible flag for this stream in the topology, but this PR will power off the iDisp link when calling hda_dsp_set_D0_state() to set D0I3 state...

@ranj063 do you follow me?

**Comment by plbossart** [comment]
> Imagine that if we want to run HDMI audio on D0i3 and specify d0i3_compatible flag for this stream in the topology, but this PR will power off the iDisp link when calling hda_dsp_set_D0_state() to set D0I3 state...
> 
> @ranj063 do you follow me?

if you specify that the stream is D0i3 compatible then it's got to be supported in hardware. If you put crazy values in the topology you will get random behavior. The only ask at this point is that the values in the topology don't result in a system crash, but there is no requirement that the system be functional when you ask it to do silly things.

**Comment by ranj063** [comment]
> > > Hi @ranj063 @kv2019i @plbossart only one concern here, is our Audio IP designed to run iDisp link with D0i3 set? If so, maybe not wise to power down dsp link blindly here, otherwise, we might need revisit this part and back and forth change could happen.
> > 
> > 
> > D0i3 is opt-in, we have to specify which streams/outputs are D0i3 compatible so we are in complete control here.
> 
> Imagine that if we want to run HDMI audio on D0i3 and specify d0i3_compatible flag for this stream in the topology, but this PR will power off the iDisp link when calling hda_dsp_set_D0_state() to set D0I3 state...
> 
> @ranj063 do you follow me?

@keyonjie no it wont. We also make sure to check that the hdac bus is powered off before turning off the link.

**Comment by ranj063** [comment]
> > > > Hi @ranj063 @kv2019i @plbossart only one concern here, is our Audio IP designed to run iDisp link with D0i3 set? If so, maybe not wise to power down dsp link blindly here, otherwise, we might need revisit this part and back and forth change could happen.
> > > 
> > > 
> > > D0i3 is opt-in, we have to specify which streams/outputs are D0i3 compatible so we are in complete control here.
> > 
> > 
> > Imagine that if we want to run HDMI audio on D0i3 and specify d0i3_compatible flag for this stream in the topology, but this PR will power off the iDisp link when calling hda_dsp_set_D0_state() to set D0I3 state...
> > @ranj063 do you follow me?
> 
> @keyonjie no it wont. We also make sure to check that the hdac bus is powered off before turning off the link.

I mean all all codecs on the bus are powered off.

**Comment by kv2019i** [comment]
@keyonjie I think this is ok. The patch does check whether iDisp codec is powered and will not disable the link if it is.

If we have a D0i3-compatible iDisp stream, the codec would have to be powered, so no action would be taken by this new code. DSP could enter D0i3, but HDA controller, link and the codec would all remain powered.

**Comment by keyonjie** [comment]
> @keyonjie I think this is ok. The patch does check whether iDisp codec is powered and will not disable the link if it is.
> 
> If we have a D0i3-compatible iDisp stream, the codec would have to be powered, so no action would be taken by this new code. DSP could enter D0i3, but HDA controller, link and the codec would all remain powered.

Oh, I didn't see this, then I am good, thanks.

**Review by keyonjie — APPROVED** [review]

**Comment by kv2019i** [comment]
Doh, @ranj063 and @keyonjie I just realized we could greatly simplify this patch when https://github.com/thesofproject/linux/pull/2732 . If iDisp codec is suspended, PR2732 will disable both link and DMA. So only thing left to do for D0i3.

Let me work on this today.

**Comment by ranj063** [comment]
> Doh, @ranj063 and @keyonjie I just realized we could greatly simplify this patch when #2732 . If iDisp codec is suspended, PR2732 will disable both link and DMA. So only thing left to do for D0i3.
> 
> Let me work on this today.

@kv2019i I thought about this but I left it nevertheless because it simplifies the flow for S0 and S0ix

**Pull request merged**